### PR TITLE
[Congestion model] Implement SoftBackpressure strategy.

### PR DESCRIPTION
**Pull request by robin-near**
at this point I don't know what I'm doing anymore.

It has no queues. It currently achieves this:

```
WORKLOAD                 STRATEGY                                 BURNT GAS    TRANSACTIONS FINISHED MEDIAN TX DELAY   MAX QUEUE LEN  MAX QUEUE SIZE  MAX QUEUE PGAS
Balanced                 Soft backpressure                        4049 PGas                    38200             308             417        247.4 KB              70
Increasing Size          Soft backpressure                        4024 PGas                    37971             204             657        256.0 MB              45
Extreme Increasing Size  Soft backpressure                        3876 PGas                    36055             218            1450          1.4 GB              36
Shard War                Soft backpressure                        3892 PGas                    31353             256           12896        603.6 MB              81
All To One               Soft backpressure                        1046 PGas                     9894             380             272        251.6 KB              70
Indirect All To One      Soft backpressure                        1157 PGas                     9782             327            1122        912.4 KB             246
Linear Imbalance         Soft backpressure                        3905 PGas                   318152              84            1731        672.8 KB              63
Big Linear Imbalance     Soft backpressure                        3870 PGas                   533304             214            2931          4.2 GB              10
```

### CODE REVIEW

**Review by jakmeier — APPROVED** [review]
Thank you for putting this idea into code and providing the summary table! @robin-near 

I can't say I understand in details how the tx limits are computed exactly, let alone the motivations behind it. :see_no_evil: A sentence or two in plain english describing the main ideas would go a long way here. :wink:

But it looks like you measure how much outstanding attached gas we have globally in the system, you attribute it to each shard that accepted the work in the first place, and then you somehow limit how much each shard is allowed to take in this round based on how much they contributed to the work already in the system. Is this a fair high-level overview?

If this is right, this seems like a very sound way to address the most fundamental problem, namely the problem of limiting total inflow to total outflow. Also, the approach to only look at which shard a transaction was originally accepted in, rather than which shard sent the receipt, is completely novel compared to strategies we evaluated so far.

The main concern I have is that it's not a very localized algorithm. We have to share a lot of information between shards. I was hoping we can find a more scalable algorithm. But it's probably not a big issue.

Also, I wonder how we could argue to find any sort of (soft or hard) limit for a single queue in the system. Our main goal for the project was to limit how much memory a validator needs for receipts. But I failed to find such an argument for this strategy. As far as I can tell, the algorithm does not even attempt to limit the length of a single delayed receipts queue, right? In theory, all globally outstanding receipts could be waiting in the same delayed receipts queue, right? I could be completely off, of course. So please help me understand this better! 

Overall, I would love to merge this soon, so I can also play around with it and see how it performs. Documentation can follow later.

File: tools/congestion-model/src/strategy/soft_backpressure.rs [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        for tx in orig_tx_ids {
+            let tx_sender = ctx.tx_sender(tx);
+            let tx_receiver = ctx.tx_receiver(tx);
+            let tx_gas = ctx.tx_attached_gas(tx);
+            *orig_rec_txn_input_gas.entry((tx_sender, tx_receiver)).or_default() += tx_gas;
+        }
+
+        let incoming_receipt_gas: GGas =
+            ctx.incoming_receipts().iter().map(|receipt| receipt.attached_gas).sum();
+        let prev_gas_burnt = ctx.gas_burnt();
```
[the remark this answers is not in the document]
Comment (robin-near) [reply tools/congestion-model/src/strategy/soft_backpressure.rs:83]: Ah yes I meant tx_gas_burnt.

File: tools/congestion-model/src/strategy/soft_backpressure.rs [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        }
+
+        let mut orig_rec_txn_input_gas = HashMap::new();
+        let orig_tx_ids = ctx
+            .incoming_receipts()
+            .iter()
+            .map(|receipt| receipt.transaction_id())
+            .collect::<Vec<_>>();
+        for tx in orig_tx_ids {
+            let tx_sender = ctx.tx_sender(tx);
```
[the remark this answers is not in the document]
Comment (robin-near) [reply tools/congestion-model/src/strategy/soft_backpressure.rs:75]: Yeah, I originally was thinking of "which shard forwarded this receipt to us" but gravitated towards going all the way to the source, as that makes more sense in this strategy where the only thing we can do is limit the source. I agree that it can be useful in other strategies as well. One caveat is that it's not very easy to define the exact behavior we want here. What I've implemented here is an approximation; the more accurate technique would be to start from each txn and look at the aggregate computation it induces in each shard, but that requires a lot more bookkeeping, and since txns take different lengths in time, it's not clear how to define a recent window to use for estimation.

File: tools/congestion-model/src/model/chunk_execution.rs [hunk 3]
```diff
@@ -135,4 +135,8 @@ impl<'model> ChunkExecutionContext<'model> {
     pub fn tx_receiver(&self, id: TransactionId) -> ShardId {
         self.transactions[id].initial_receipt_receiver()
     }
+
+    pub fn tx_sender(&self, id: TransactionId) -> ShardId {
+        self.transactions[id].sender_shard
```
[the remark this answers is not in the document]
Comment (robin-near) [reply tools/congestion-model/src/model/chunk_execution.rs:140]: Oh, I see, yeah... this data is on a different shard. That does seem like a non-trivial problem.

File: tools/congestion-model/src/strategy/soft_backpressure.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+#[derive(Default)]
+pub struct SoftBackpressure {
+    shard_id: Option<ShardId>,
+    all_shards: Vec<ShardId>,
+}
+
+#[derive(Default, Clone, Debug)]
+struct ShardState {
+    orig_rec_txn_input_gas: HashMap<(ShardId, ShardId), GGas>,
```
[the remark this answers is not in the document]
Comment (robin-near) [reply tools/congestion-model/src/strategy/soft_backpressure.rs:14]: I'm also not very fond of the (shard x shard) approach, as beyond a few shards this explodes quickly, and limiting a high dimensional input is unlikely to be effective. For example if we limit (56, 80) because 80 is overloaded and we saw a lot of traffic coming from (56, 80), next block we may see a (61, 80). But if we limit all of (x, 80) then because we make an independent decision on x (source shard), we'd have to apply a very low limit and end up not being able to take in any transactions at all. I'm not sure what to do.
